text/0507-subnets.md

```diff
@@ -0,0 +1,402 @@
+# Full control over VPC and subnet configuration
+
+* **Original Author(s):**: @otaviomacedo
+* **Tracking Issue**: #507
+* **API Bar Raiser**: @{BAR_RAISER_USER}
+
+New L2 constructs to allow for greater control over VPC design. Among other
+things, these constructs allow users to create and configure subnets in
+ways that are not currently possible with the `Vpc` construct.
+
+## Working Backwards
+
+### IP addressing
+
+When you create a VPC, in addition to the primary IP block, you can provide
+one or more secondary IP blocks, by providing them in the CIDR format:
+
+```ts
+const vpc = new Vpc(this, 'vpc', {
+  // Primary address block
+  ipAddresses: IpAddresses.ipv4('10.1.0.0/16'),
+});
+
+// The secondary address block must be in the same RFC 1918 range
+vpc.addSecondaryAddressBlock(IpAddresses.ipv4('10.2.0.0/16'));
+```
+
+Or by providing an [IPAM] pool with a netmask length:
+
+```ts
+const ipam = new Ipam(stack, 'ipam');
+const pool = ipam.publicScope.addPool({
+  addressFamily: AddressFamily.IP_V4,
+  provisionedCidrs: [IpAddresses.ipv4('10.2.0.0/16')],
+});
+
+vpc.addSecondaryAddressBlock(IpAddresses.ipv4Ipam({
+  ipamPool: pool,
+  netmaskLength: 20
+}));
+```
+
+You can also add secondary IPv6 address blocks, in three different ways:
+
+```ts
+// 1. Providing an Ipv6 address block. Because IPv6 addresses are all publicly 
+// addressable, they must come from an address pool that you own and brought to 
+// AWS (BYOIP). So you must also provide the pool ID:
+vpc.addSecondaryAddressBlock(IpAddresses.ipv6({
+  cidr: '2001:db8:1234:1a00::/56',
+
+  // The pool of IPs you own. Not to be confused with an IPAM pool ID
+  poolId: 'my-ipv6-pool-id',
+}));
+
+// 2. Using an IPAM pool:
+vpc.addSecondaryAddressBlock(IpAddresses.ipv6Ipam({
+  ipamPoolId: pool,
+  netmaskLength: 64
+}));
+
+// 3. Using an Amazon-provided IPv6 CIDR block:
+vpc.addSecondaryAddressBlock(IpAddresses.amazonProvidedIpv6());
+```
+
+### Defining your own subnets
+
+By default, `Vpc` will create subnets for you (by default, a public and a
+private subnet per availability zone). If you want to define your own
+subnets, use the `addSubnet()` method, which will turn off this default
+behavior:
```

Comment:
Is the plan to lazily create the default subnets when `selectSubnets` is called, if `addSubnet` hasn't been called at that point yet? (And then disallow `addSubnet` afterwards?)

Reply:
I was thinking of doing it the other way around: let the default subnets be created eagerly (as it is today). But the first `addSubnet` call will override them. Are you concerned about `selectSubnets` returning different values at different points in time?